semgrep-core/src/fixing/Autofix_metavar_replacement.ml

```diff
@@ -0,0 +1,125 @@
+  tbl
+
+let replace metavar_tbl pattern_ast =
+  (* Use a mapper to traverse the AST. For each metavar, look up what it is
+   * bound to. If the kind of node matches, replace the metavar in the pattern
+   * with the AST node to which it is bound. Note that we can't handle this by
+   * simply visiting the identifier nodes, because metavars are often bound to
+   * nodes other than identifiers. Instead, we have to inspect certain kinds of
+   * nodes for metavars and then replace the entire node. *)
+  let mapper =
+    Map_AST.(
+      mk_visitor
+        {
+          default_visitor with
+          (* TODO handle more kinds of nodes *)
+          kargs =
+            (fun (k, _) args ->
+              (* A metavariable can appear as a single argument, but can be
+               * bound to multiple arguments, so we have to handle this case by
+               * looking at an argument list as a whole *)
+              let map_arg arg =
+                match arg with
+                | Arg { e = N (Id ((id_str, _), _)); _ } -> (
+                    match Hashtbl.find_opt metavar_tbl id_str with
+                    | Some (MV.Args args) -> args
+                    | _ -> [ arg ])
+                | _ -> [ arg ]
+              in
+              let args = List.concat_map map_arg args in
+              k args);
+          kexpr =
+            (fun (k, _) orig ->
+              (* The mapper changes the ID of expressions. This stymies the
+               * print-avoidance implemented later in the autofix process,
+               * because the resulting AST nodes cannot be recognized as the
+               * same as the original ones.
+               *
+               * To work around this, we use the original expression, except for
+               * the mapped kind.
+               *
+               * There is a note in the mapper expressing uncertainty about
+               * whether the ID should be reused or not. We should consider
+               * changing that behavior in the mapper, in which case this
+               * workaround would no longer be necessary.
+               * *)
+              let expr = k orig in
+              { orig with e = expr.e });
+        })
+  in
+  mapper.Map_AST.vany pattern_ast
+
+(* Check for remaining metavars in the AST. If there are any, that indicates a
+ * failure to properly replace them in the previous step, and the autofix
+ * attempt should be aborted.
+ *
+ * This currently works by comparing all identifiers against the list of bound
+ * metavariables. Some languages allow identifiers that look like metavariables,
+ * so this is less likely to result in a false positive than if we checked for
+ * any identifiers that could be metavariables. However, this wouldn't catch the
+ * case of a malformed fix pattern where the user wrote a metavariable in the
+ * fix that doesn't exist in the rule's pattern.
+ * *)
+let has_remaining_metavars metavar_tbl ast =
```

Comment:
Why not just perform this check when mapper hits the `Id` case? Anything that could have been resolved would be resolved by then. And we wouldn't have to go to the expense of creating another visitor and traversing the AST again

Reply:
I guess it's not so bad since fix patterns are small, so feel free to ignore this

Reply:
I think this is more foolproof because a maintainer of the mapper could forget to call the `k` function to continue the traversal, potentially leading to missed nodes.